Updated Argon2 Support

kvnallsn opened the pull request:
Based off the work @ccfb3ee765a58cae did for Argon2 support, but updated to work with the latest changes on the master branch.

This exposes libsodium's crypt_pwhash_argon2 functions, providing a rust API similar to sodiumoxide::crypto::pwhash::scryptsalsa208sha256::*

nvesely commented:
Hey, I forgot about this for a while, feel free to ping me in the future. Will get to a review soon.

Dylan-DPC commented:
From the meeting: 
> A cursory overview reviews some mistakes, e.g. a mutable pointer cast from a const pointer to a non-mutable object

kpp commented:
> Hey, I forgot about this for a while, feel free to ping me in the future. Will get to a review soon.

@nvesely ping

kvnallsn commented:
@nvesely I've redesigned this module to be better in line with what you all are looking for I think.  Let me know what more changes need to be done.

So far:

- Fixed mut pointer to const structure
- Added tests from upstream libsodium
- Split into two different modules
- Added blurbs about differences between Argon2i and Argon2id

kpp commented:
> continuous-integration/appveyor/branch Expected — Waiting for status to be reported

Looks like AppveyorCI does not report the status sometimes)

kpp commented on src/crypto/pwhash/mod.rs:
```diff
@@ -65,4 +65,8 @@
+mod argon2_macros;
+pub mod argon2i13;
+pub mod argon2id13;
+pub mod scryptsalsa208sha256;
```
Add newline at EOF

kpp commented on src/crypto/pwhash/argon2_macros.rs:
```diff
@@ -0,0 +1,197 @@
+    }
+}
+
+));
```
Add newline at EOF

kpp commented on src/crypto/pwhash/argon2_macros.rs:
```diff
@@ -0,0 +1,197 @@
+    OpsLimit(opslimit): OpsLimit,
+    MemLimit(memlimit): MemLimit,
+) -> Result<&'a [u8], ()> {
+    if unsafe {
```
`if unsafe {}` -> `let res = unsafe {}; if res ...`

kpp commented on src/crypto/pwhash/argon2_macros.rs:
```diff
@@ -0,0 +1,197 @@
+    MemLimit(memlimit): MemLimit,
+) -> Result<HashedPassword, ()> {
+    let mut _out = HashedPassword([0; HASHEDPASSWORDBYTES]);
+    if unsafe {
```
`if unsafe {}` -> `let res = unsafe {}; if res ...`

kpp commented on src/crypto/pwhash/argon2id13.rs:
@kvnallsn where did you take these vectors?

kvnallsn replied:
These are from libsodium itself, from test/default/pwhash_argon2*

kvnallsn replied:
Initially I had an implementation of hex2bin and bin2hex but removed them in favor of using the raw vectors and wait for #252 to land.  I don't see much reason to change this until that patch gets merged. Even then, I think the raw vectors fit well here since the test methodology is different. Libsodium does all tests then compares the resulting output to a reference file whereas we can compare results individually.

kpp commented on src/crypto/pwhash/argon2_macros.rs:
```diff
@@ -0,0 +1,197 @@
+    let mut salt = Salt([0; SALTBYTES]);
+    {
+        let Salt(ref mut sb) = salt;
+        randombytes_into(sb);
```
how about `randombytes_into(salt.0)?

kvnallsn replied:
Yep, meant to change that.  Fixed now

kpp commented on src/crypto/pwhash/argon2_macros.rs:
```diff
@@ -0,0 +1,197 @@
+    OpsLimit(opslimit): OpsLimit,
+    MemLimit(memlimit): MemLimit,
+) -> Result<HashedPassword, ()> {
+    let mut _out = HashedPassword([0; HASHEDPASSWORDBYTES]);
```
why is it called `_out`?

kvnallsn replied:
To suppress an unused mut warning.  I believe `out` needs to be mutable due to borrowing as a mutable pointer in the unsafe call. Will change this if not, or if there's a better way to do it

samscott89 replied:
Try `out.0.as_mut_ptr()`[:)](https://doc.rust-lang.org/std/primitive.slice.html#method.as_mut_ptr)

kpp commented on src/crypto/pwhash/argon2_macros.rs:
```diff
@@ -0,0 +1,197 @@
+///
+/// It returns `true` if the verification succeeds, and `false` on error.
+pub fn pwhash_verify(hp: &HashedPassword, passwd: &[u8]) -> bool {
+    unsafe {
```
`unsafe {} == 0` -> `let ret = unsafe {}; ret == 0`

Dylan-DPC commented:
bors: r+
